### ISSUE TITLE
Incorrect warning for unused field.

### ISSUE THREAD

**Issue by rcls** [rust-lang/rust#92995]
(Note entirely sure what the intended behaviour is here, but named fields versus tuple fields are treated differently, which seems wrong.)

Given the following code: (https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8c1b3d36eae323990f50c01d182c042d)

```rust
#[derive(Debug)]
pub struct Foo{x:i32}

fn main() {
    let foo = Foo{x:1};
    println!("{:?}", foo);
}
```

The current output is:

```
warning: field is never read: `x`
```
But the field is actually read due to the `#[derive(Default)]` and the use of that.

Arguable, a warning here is sensible in that as the usage is "debug" output only. But in that case it is hard to see why this doesn't warn also (https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a03f26d9f9c53bda612313913bfaf67e):
```rust
#[derive(Debug)]
pub struct Foo(i32);

fn main() {
    let foo = Foo(1);
    println!("{:?}", foo);
}
```
At least these two should be treated consistently?

**Comment by ehuss** [comment]
Thanks for the report!  This is currently intended behavior, as otherwise adding `#[derive(Debug)]` would prevent the unused warning from working. I understand it can be problematic as actually using the debug formatting seems like it should count as a "use".  There is a deeper explanation at https://github.com/rust-lang/rust/issues/88900#issuecomment-929493011.

Closing as a duplicate of #88900, and the tuple issue is #92790.
